### PR TITLE
Feature/gem 189 password recovery

### PR DESCRIPTION
**Pull request by maltem-za**

### CODE REVIEW

**Comment by maltem-za** [comment]
@miltontony Please review :)
Note that there will be two 0004 migrations after merging. Not sure if that'll be an issue, I'd think not as they're also date and timestamped. Happy to rename the one in this PR to 0005 if desired.

File: gem/views.py [hunk 1]
```diff
@@ -55,14 +64,172 @@ def form_valid(self, form):
...
+
+        username = form.cleaned_data['username']
+        random_security_question_idx = self.request.session[
+            'random_security_question_idx'
+        ]
+        random_security_question_answer = form.cleaned_data[
+            'random_security_question_answer'
+        ]
+
+        # TODO: consider moving these checks to GemForgotPasswordForm.clean()
```
Comment (miltontony) [inline gem/views.py]: Agreed. Happy for you to do this in a separate PR

File: gem/views.py [hunk 2]
```diff
@@ -55,14 +64,172 @@ def form_valid(self, form):
...
+                           'Your answer to the security question was invalid. '
+                           'Please try again.')
+            return self.render_to_response({'form': form})
+
+        # NB: NOT safe if cookie-based sessions are used (with cookie-based
+        # sessions the session data is stored in the cookie itself and is
+        # NOT encrypted!)
+        # See https://docs.djangoproject.com/en/1.9/topics/http/sessions/
+        # #configuring-the-session-engine
+        # TODO: consider generating a reset token for the user and redirecting
```
Comment (miltontony) [inline gem/views.py]: Agreed. Please use token and redirect instead. See this example from [django email reset](https://github.com/django/django/blob/master/django/contrib/auth/forms.py#L274)

File: gem/views.py [hunk 3]
```diff
@@ -55,14 +64,172 @@ def form_valid(self, form):
...
+
+        if not user.is_active:
+            return HttpResponseForbidden()
+
+        password = form.cleaned_data['password']
+        confirm_password = form.cleaned_data['confirm_password']
+
+        if password != confirm_password:
+            form.add_error('password',
+                           'The two PINs that you entered do not match. '
```
Comment (miltontony) [inline gem/views.py]: This needs to be wrapped in a `_(..)` for translations

File: gem/views.py [hunk 4]
```diff
@@ -55,14 +64,172 @@ def form_valid(self, form):
...
+                user.gem_profile.check_security_question_2_answer(
+                    random_security_question_answer
+                )
+        else:
+            logging.warn('Unhandled security question index')
+
+        if not is_answer_correct:
+            self.request.session['forgot_password_attempts'] += 1
+            form.add_error('random_security_question_answer',
+                           'Your answer to the security question was invalid. '
```
Comment (miltontony) [inline gem/views.py]: This needs to be wrapped in a `_(..)` for translations

File: gem/views.py [hunk 5]
```diff
@@ -55,14 +64,172 @@ def form_valid(self, form):
...
+        ]
+
+        # TODO: consider moving these checks to GemForgotPasswordForm.clean()
+        # see django.contrib.auth.forms.AuthenticationForm for reference
+        try:
+            user = User.objects.get_by_natural_key(username)
+        except User.DoesNotExist:
+            self.request.session['forgot_password_attempts'] += 1
+            form.add_error(
+                'username', 'The username that you entered appears to be '
```
Comment (miltontony) [inline gem/views.py]: This needs to be wrapped in a `_(..)` for translations

File: gem/views.py [hunk 6]
```diff
@@ -55,14 +64,172 @@ def form_valid(self, form):
...
+            self.request.session['forgot_password_attempts'] = 0
+
+        if self.request.session['forgot_password_attempts'] >= 5:
+            # GEM-195 implemented a 10 min session timeout, so effectively
+            # the user can only try again once their anonymous session expires.
+            # If they make another request within the 10 min time window the
+            # expiration will be reset to 10 mins in the future.
+            # This is obviously not bulletproof as an attacker could simply
+            # not send the session cookie to circumvent this.
+            form.add_error(None, 'Too many attempts. Please try again later.')
```
Comment (miltontony) [inline gem/views.py]: This needs to be wrapped in a `_(..)` for translations

File: gem/templates/reset_password.html [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                {% endif %}
+
+                <p><strong>{% trans "PIN" %}</strong></p>
+                {{ form.password }}
+
+                <p><strong>{% trans "Confirm PIN" %}</strong></p>
+                {{ form.confirm_password }}
+
+                <br>
+                <input type="submit" value='{% trans "RESET PIN" %}' />
```
Comment (miltontony) [inline gem/templates/reset_password.html]: I think the case for "Reset PIN" needs to match `line 7` and any uppercasing should be added in the CSS if needed.

File: gem/templates/forgot_password.html [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+                {{ form.username }}
+
+                <p><strong>{% trans "Security question" %}</strong></p>
+                <p>{% trans "Please answer the following security question." %}</p>
+
+                <p><strong>{% trans random_security_question %}</strong></p>
+                {{ form.random_security_question_answer }}
+
+                <br>
+                <input type="submit" value='{% trans "FORGOT PIN" %}' />
```
Comment (miltontony) [inline gem/templates/forgot_password.html]: I think "FORGOT" needs to be title case to match `line 7`. Any upper casing should be handled in the CSS if needed.

File: gem/views.py (hunk 3 above)
Comment (maltem-za) [inline gem/views.py]: Done

File: gem/views.py (hunk 4 above)
Comment (maltem-za) [inline gem/views.py]: Done

File: gem/views.py (hunk 5 above)
Comment (maltem-za) [inline gem/views.py]: Done

File: gem/views.py (hunk 6 above)
Comment (maltem-za) [inline gem/views.py]: Done

File: gem/templates/reset_password.html (hunk 7 above)
Comment (maltem-za) [inline gem/templates/reset_password.html]: Done

File: gem/templates/forgot_password.html (hunk 8 above)
Comment (maltem-za) [inline gem/templates/forgot_password.html]: Done

**Comment by miltontony** [comment]
@maltem-za will you be making 2 PRs as follow ups to the 2 TODO's raised here?

File: gem/views.py (hunk 2 above)
Comment (maltem-za) [inline gem/views.py]: Done

**Comment by maltem-za** [comment]
@miltontony I've done the token thing, other outstanding TODO will have to be done some other time

**Comment by miltontony** [comment]
:+1:  thanks!
